### ISSUE TITLE
Unable to narrow a union of classes (not instances)

### ISSUE THREAD

**Issue by psychedelicious** [microsoft/pyright#7393]
**Describe the bug**

I'm unable to narrow a union of classes with the `is` operator. This works at runtime, though maybe it isn't technically valid python?

**Code or Screenshots**

In the example, pyright doesn't narrow the type of `t` to `type[Foo]`:

```python
class Foo:
    def __init__(self, x: int):
        self.x = x
        pass

    pass


class Bar:
    def __init__(self) -> None:
        pass


def get_type() -> type[Foo] | type[Bar]:
    return Bar


t = get_type()

if t is Foo:
    # I expect `t` to be narrowed to `type[Foo]`, but pyright shows `(variable) t: type[Foo] | type[Bar]`
    t(x=1)  # No parameter named "x"PylancereportCallIssue
    print("Got a Foo")
```

Pyright output:
```
pyright narrow.py
.../narrow.py
  .../narrow.py:22:7 - error: No parameter named "x" (reportCallIssue)
1 error, 0 warnings, 0 informations 
```

**VS Code extension or command-line**
VSCode extension and CLI:
```
pyright --version
pyright 1.1.352
```

Possibly related to  #7214

**Comment by erictraut** [comment]
This isn't a supported type guard. For a complete list of supported forms, refer to [this documentation](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=type-guards). 

We do occasionally add support for new type guard forms, but each one requires specialized logic and can affect general performance of the type analyzer, so we do so only in cases where there's strong demand or evidence that it's a common pattern. The pattern you're using here is not something I've seen used previously, so for now it's not a type guard pattern we will add. Incidentally, it's also not supported by other type checkers by mypy, which further supports the conclusion that it's not a common pattern.

I am willing to revisit this decision in the future if there is strong evidence (in the form of upvotes), but for now I'm going to close this issue since I don't have any plans to implement this functionality.

**Comment by psychedelicious** [comment]
Thanks for the prompt and thorough response, @erictraut. I'll remember to check the linked documentation in the future.
